--- pysimplesoap/server.py~	2013-01-14 10:05:04.000000000 -0700
+++ pysimplesoap/server.py	2013-01-16 11:58:34.000000000 -0700
@@ -126,7 +126,8 @@
             # detect soap prefix and uri (xmlns attributes of Envelope)
             for k, v in request[:]:
                 if v in ("http://schemas.xmlsoap.org/soap/envelope/",
-                                  "http://www.w3.org/2003/05/soap-env",):
+                                  "http://www.w3.org/2003/05/soap-env",
+                                  "http://www.w3.org/2003/05/soap-envelope",):
                     soap_ns = request.attributes()[k].localName
                     soap_uri = request.attributes()[k].value
                 
